code/game/objects/items/devices/radio/radio.dm

```diff
@@ -45,6 +45,51 @@
 	var/list/secure_radio_connections
 	var/radio_silent = FALSE // If true, radio doesn't make sound effects (ie for Syndicate internal radio implants)
 
+	/// this is necessary to adjust 'known_channels' in mob var properly
+	var/mob/current_holder = null
+
+/obj/item/radio/Initialize(mapload)
+	wires = new /datum/wires/radio(src)
+	if(prison_radio)
+		wires.cut(WIRE_TX) // OH GOD WHY
+	secure_radio_connections = new
+	. = ..()
+	frequency = sanitize_frequency(frequency, freerange)
+	set_frequency(frequency)
+
+	for(var/ch_name in channels)
+		secure_radio_connections[ch_name] = add_radio(src, GLOB.radiochannels[ch_name])
+
+	become_hearing_sensitive(ROUNDSTART_TRAIT)
+
+/obj/item/radio/ComponentInitialize()
+	. = ..()
+	AddComponent(/datum/component/empprotection, EMP_PROTECT_WIRES)
+
+/obj/item/radio/Destroy()
+	remove_radio_all(src) //Just to be sure
+	QDEL_NULL(wires)
+	QDEL_NULL(keyslot)
+	return ..()
+
+/obj/item/radio/pickup(mob/user)
+	. = ..()
+	current_holder = user
+	if(current_holder)
+		addtimer(CALLBACK(current_holder, /mob/.proc/refresh_known_radio_channels), 2)
+
+/obj/item/radio/dropped(mob/user, silent)
+	. = ..()
+	if(current_holder)
+		addtimer(CALLBACK(current_holder, /mob/.proc/refresh_known_radio_channels), 2)
+	current_holder = null
+
+/obj/item/radio/equipped(mob/user, slot)
+	. = ..()
+	current_holder = user
+	if(current_holder)
+		addtimer(CALLBACK(current_holder, /mob/.proc/refresh_known_radio_channels), 2)
```

Comment:
This 0.2 second delay feels incredibly janky, would it be better to main the list of known radio channels and add new ones when a radio is equipped to the correct slot?

For example, when a security headset is equipped to the ear slot, add 1 to RADIO_CHANNEL_SECURITY.
When a player picks up a security radio add 1 to RADIO_CHANNEL_SECURITY.
When the player takes off their headset remove 1.
If the player drops their security radio, the count is now 0 so they can no longer hear it.

This would also avoid the pretty expensive refresh_known_radio_channels() along with all its snowflake behaviour

Reply:
It is related to how `Destroy()` proc works. It should be called after destroyed, but the game still thinks they can access to the channel, so I made it like that.(I mean, drop/destroyed is called later after a point I want it to happen) I am doubtful that your suggestion wouldn't change that weird behaviour, and I have no clue how I can fix that without using addtimer.

Reply:
Well, let me test something then I will tell you

Reply:
I think changing it to 'add/remove' logic would make it any good. It's because there are not a few things to check
* checks Listening
* checks if it is headset, and be at your ears
* checks if it is in your bag
* checks keyslot
* checks keyslot2
* and so on
and these things are not only single checkable. I mean, these are even should be checked together, which means it needs incredibly a lot of condition checks.
but also how `pickup/dropped/equipped` procs work makes my brain explode.

Reply:
`destory()` calls `dropped()`, but not at the moment I want. That's why I had to put 0.2s delay.